src/components/tl/mlx5/mcast/tl_mlx5_mcast_helper.c

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_mlx5_mcast_helper.h"
+#include <glob.h>
+#include <net/if.h>
+#include <ifaddrs.h>
+
+#define PREF        "/sys/class/net/"
+#define SUFF        "/device/resource"
+#define MAX_STR_LEN 128
+
+static ucc_status_t ucc_tl_get_ipoib_ip(char *ifname, struct sockaddr_storage *addr)
+{
+    ucc_status_t    status  = UCC_ERR_NO_RESOURCE;
+    struct ifaddrs *ifaddr  = NULL;
+    struct ifaddrs *ifa     = NULL;
+    int             is_ipv4 = 0;
+    int             family;
+    int             n;
+    int             is_up;
+
+    if (getifaddrs(&ifaddr) == -1) {
+        return UCC_ERR_NO_RESOURCE;
+    }
+
+    for (ifa = ifaddr, n = 0; ifa != NULL; ifa=ifa->ifa_next, n++) {
+        if (ifa->ifa_addr == NULL) {
+            continue;
+        }
+
+        family = ifa->ifa_addr->sa_family;
+        if (family != AF_INET && family != AF_INET6) {
+            continue;
+        }
+
+        is_up   = (ifa->ifa_flags & IFF_UP) == IFF_UP;
+        is_ipv4 = (family == AF_INET) ? 1 : 0;
+
+        if (is_up && !strncmp(ifa->ifa_name, ifname, strlen(ifname)) ) {
+            if (is_ipv4) {
+                memcpy((struct sockaddr_in *) addr,
+                       (struct sockaddr_in *) ifa->ifa_addr,
+                       sizeof(struct sockaddr_in));
+            } else {
+                memcpy((struct sockaddr_in6 *) addr,
+                       (struct sockaddr_in6 *) ifa->ifa_addr,
+                       sizeof(struct sockaddr_in6));
+            }
+
+            status = UCC_OK;
+            break;
+        }
+    }
+
+    freeifaddrs(ifaddr);
+    return status;
+}
+
+static int cmp_files(char *f1, char *f2)
+{
+    int   answer = 0;
+    FILE *fp1;
+    FILE *fp2;
+    int   ch1;
+    int   ch2;
+
+    if ((fp1 = fopen(f1, "r")) == NULL) {
+        goto out;
+    } else if ((fp2 = fopen(f2, "r")) == NULL) {
+        goto close;
+    }
+
+    do {
+        ch1 = getc(fp1);
+        ch2 = getc(fp2);
+    } while((ch1 != EOF) && (ch2 != EOF) && (ch1 == ch2));
+
+
+    if (ch1 == ch2) {
+        answer = 1;
+    }
+
+    fclose(fp2);
+close:
+    fclose(fp1);
+out:
+    return answer;
+}
+
+static int port_from_file(char *port_file)
+{
+    int   res = -1;
+    char  buf1[MAX_STR_LEN];
+    char  buf2[MAX_STR_LEN];
+    FILE *fp;
+    int   len;
+
+    if ((fp = fopen(port_file, "r")) == NULL) {
+        return -1;
+    }
+
+    if (fgets(buf1, MAX_STR_LEN - 1, fp) == NULL) {
+        goto out;
+    }
+
+    len       = strlen(buf1) - 2;
+    strncpy(buf2, buf1 + 2, len);
```

Comment:
why do we need to skip the first two characters here?

Reply:
I am not sure why, these parts are from the legacy HCOLL code that I borrowed. They work as intended so I did not bother to modify it